File: C210/ga-java/src/br/inatel/c210/ga/algorithm/GeneticUtils.java
package br.inatel.c210.ga.algorithm;

import java.util.List;
import java.util.Random;

import br.inatel.c210.ga.entity.Chromossome;
import br.inatel.c210.ga.entity.Problem;

public class GeneticUtils
{
    public static final Random RANDOM = new Random(42L);

    public static Chromossome findBestChromossome(List<Chromossome> population)
    {
        return population.get(findBestChromossomeIndex(population));
    }

    public static int findBestChromossomeIndex(List<Chromossome> population)
    {
        double maxScore = Double.MIN_VALUE;
        int maxIndex = -1;
        for (int i = 0; i < population.size(); i++)
        {
            double score = Problem.g(population.get(i));
            if (score > maxScore)
            {
                maxScore = score;
                maxIndex = i;
            }
        }
        return maxIndex;
    }

    public static Chromossome findWorstChromossome(List<Chromossome> population)
    {
        return population.get(findWorstChromossomeIndex(population));
    }

    public static int findWorstChromossomeIndex(List<Chromossome> population)
    {
        double minScore = Double.MAX_VALUE;
        int minIndex = -1;
        for (int i = 0; i < population.size(); i++)
        {
            double score = Problem.g(population.get(i));
            if (score < minScore)
            {
                minScore = score;
                minIndex = i;
            }
        }
        return minIndex;
    }

    public static String formatChromossome(Chromossome c)
    {
        return String.format("%s, Score = %.3f", c, Problem.g(c));
    }
}

File: C210/ga-python/modules/genetic_utils.py
from .problem import Problem

class GeneticUtils:

    @staticmethod
    def find_best_chromossome(population):
        best_chromossome = None

        for chromossome in population:
            score = Problem.g_chromossome(chromossome)

            if best_chromossome is None or score > Problem.g_chromossome(best_chromossome):
                best_chromossome = chromossome

        return best_chromossome

    @staticmethod
    def find_worst_chromossome(population):
        worst_chromossome = None

        for chromossome in population:
            score = Problem.g_chromossome(chromossome)

            if worst_chromossome is None or score < Problem.g_chromossome(worst_chromossome):
                worst_chromossome = chromossome

        return worst_chromossome

    @staticmethod
    def format_chromossome(chromossome):
        return f"{chromossome.to_string()}, Score = {Problem.g_chromossome(chromossome):.3f}"
